File: go/compress.go
package solution


func compress(input string) ([]int, []rune)  {
  indexes := make([]int, len(input))
  chars := make([]rune, 0)
  indexMap := make(map[rune]int)

  for i, c := range input {
    if indexMap[c] == 0 {
      indexMap[c] = len(indexMap) + 1
      chars = append(chars, c)
    }

    indexes[i] = indexMap[c]
  }

  return indexes, chars
}

File: rust/src/compress.rs
pub fn compress(input: &str) -> (Vec<char>, Vec<usize>) {
    let mut indexes: Vec<usize> = Vec::new();
    let mut chars: Vec<char> = Vec::new();

    input.chars().for_each(|c| {
        if !chars.contains(&c) {
            chars.push(c);
        }

        let index = chars.iter().position(|&x| x == c).expect("Index not found");

        indexes.push(index);
    });

    return (chars, indexes);
}

#[cfg(test)]
mod tests {
    use crate::compress::*;

    #[test]
    fn test_compress() {
        let result = compress("abcab");

        assert_eq!(result.0, vec!['a', 'b', 'c']);
        assert_eq!(result.1, vec![0, 1, 2, 0, 1]);
    }
}
